File: flutter-boilerplate/test/lib/subscribe_flow_test.dart
import 'package:flutter_boilerplate/lib/billing/subscribe_flow.dart';
import 'package:flutter_test/flutter_test.dart';

// BE-019: Stripe 3DS recovery on first subscribe.
void main() {
  test('a plain success passes straight through', () async {
    var confirmed = 0;
    final result = await completeSubscribeWithAuthentication(
      {'success': true, 'periodEnd': 'x'},
      confirm: (_) async => confirmed++,
      finalize: (_) async => throw StateError('must not finalize'),
    );
    expect(result['success'], isTrue);
    expect(confirmed, 0);
  });

  test('confirms the PaymentIntent, then finalizes the subscription', () async {
    final confirmedSecrets = <String>[];
    final finalized = <String>[];
    final result = await completeSubscribeWithAuthentication(
      {
        'success': false,
        'reason': 'authentication_required',
        'clientSecret': 'pi_1_secret_x',
        'stripeSubscriptionId': 'sub_1',
      },
      confirm: (s) async => confirmedSecrets.add(s),
      finalize: (id) async {
        finalized.add(id);
        return {'success': true};
      },
    );
    expect(confirmedSecrets, ['pi_1_secret_x']);
    expect(finalized, ['sub_1']);
    expect(result['success'], isTrue);
  });

  test('a declined result throws SubscribeDeclined with the backend reason',
      () async {
    expect(
      () => completeSubscribeWithAuthentication(
        {'success': false, 'reason': 'insufficient_funds'},
        confirm: (_) async {},
        finalize: (_) async => {'success': true},
      ),
      throwsA(
        isA<SubscribeDeclined>()
            .having((e) => e.reason, 'reason', 'insufficient_funds'),
      ),
    );
  });

  test('a missing success flag is a decline, never a silent success', () async {
    // Regression: the page used to treat any non-throwing subscribe() as
    // success, so a declined card showed "Upgrade successful!".
    expect(
      () => completeSubscribeWithAuthentication(
        <String, dynamic>{},
        confirm: (_) async {},
        finalize: (_) async => {'success': true},
      ),
      throwsA(isA<SubscribeDeclined>()),
    );
  });

  test('gives up when Stripe keeps asking for authentication', () async {
    final pending = {
      'success': false,
      'reason': 'authentication_required',
      'clientSecret': 'pi_1_secret_x',
      'stripeSubscriptionId': 'sub_1',
    };
    var finalizeCalls = 0;
    await expectLater(
      () => completeSubscribeWithAuthentication(
        pending,
        confirm: (_) async {},
        finalize: (_) async {
          finalizeCalls++;
          return pending;
        },
      ),
      throwsA(
        isA<SubscribeDeclined>()
            .having((e) => e.reason, 'reason', 'authentication_failed'),
      ),
    );
    // Two rounds, then it stops instead of looping on the bank forever.
    expect(finalizeCalls, 2);
  });
}

File: next-js-boilerplate/src/lib/billing/subscribe-flow.test.ts
import { describe, it, expect, vi } from "vitest";
import {
  completeSubscribeWithAuthentication,
  declineMessage,
} from "./subscribe-flow";

const t = {
  declinedInsufficientFunds: "NO_FUNDS",
  declinedCard: "DECLINED",
  authenticationRequired: "AUTH_REQUIRED",
  subscriptionFailed: "FAILED",
};

// BE-019: Stripe 3DS recovery on first subscribe.
describe("completeSubscribeWithAuthentication", () => {
  it("passes a plain success straight through", async () => {
    const confirm = vi.fn();
    const finalize = vi.fn();
    const ok = {
      ok: true,
      periodEnd: null,
      pendingTier: null,
      pendingTierEffectiveAt: null,
    };
    const result = await completeSubscribeWithAuthentication(ok, {
      confirm,
      finalize,
      authenticationFailedMessage: "AUTH_FAILED",
    });
    expect(result).toBe(ok);
    expect(confirm).not.toHaveBeenCalled();
    expect(finalize).not.toHaveBeenCalled();
  });

  it("confirms the PaymentIntent, then finalizes the subscription", async () => {
    const confirm = vi.fn().mockResolvedValue(null);
    const finalize = vi.fn().mockResolvedValue({
      ok: true,
      periodEnd: "2026-10-03",
      pendingTier: null,
      pendingTierEffectiveAt: null,
    });
    const result = await completeSubscribeWithAuthentication(
      {
        ok: false,
        periodEnd: null,
        pendingTier: null,
        pendingTierEffectiveAt: null,
        requiresAction: true,
        clientSecret: "pi_1_secret_x",
        stripeSubscriptionId: "sub_1",
      },
      { confirm, finalize, authenticationFailedMessage: "AUTH_FAILED" },
    );
    expect(confirm).toHaveBeenCalledWith("pi_1_secret_x");
    expect(finalize).toHaveBeenCalledWith("sub_1");
    expect(result.ok).toBe(true);
  });

  it("surfaces Stripe's own error when the customer fails 3DS and never finalizes", async () => {
    const confirm = vi.fn().mockResolvedValue("Your card was declined.");
    const finalize = vi.fn();
    await expect(
      completeSubscribeWithAuthentication(
        {
          ok: false,
          periodEnd: null,
          pendingTier: null,
          pendingTierEffectiveAt: null,
          requiresAction: true,
          clientSecret: "pi_1_secret_x",
          stripeSubscriptionId: "sub_1",
        },
        { confirm, finalize, authenticationFailedMessage: "AUTH_FAILED" },
      ),
    ).rejects.toThrow("Your card was declined.");
    expect(finalize).not.toHaveBeenCalled();
  });

  it("gives up with the authentication-failed message if Stripe keeps asking", async () => {
    const pending = {
      ok: false,
      periodEnd: null,
      pendingTier: null,
      pendingTierEffectiveAt: null,
      requiresAction: true,
      clientSecret: "pi_1_secret_x",
      stripeSubscriptionId: "sub_1",
    };
    const confirm = vi.fn().mockResolvedValue(null);
    const finalize = vi.fn().mockResolvedValue(pending);
    await expect(
      completeSubscribeWithAuthentication(pending, {
        confirm,
        finalize,
        authenticationFailedMessage: "AUTH_FAILED",
      }),
    ).rejects.toThrow("AUTH_FAILED");
    expect(finalize).toHaveBeenCalledTimes(2);
  });
});

describe("declineMessage", () => {
  const err = (key: string, reason?: string) =>
    Object.assign(new Error("Payment declined"), {
      exception: { key, reason },
    });

  it("maps backend reasons to readable copy instead of the raw code", () => {
    expect(
      declineMessage(
        err("billing.errors.insufficientFunds", "insufficient_funds"),
        t,
      ),
    ).toBe("NO_FUNDS");
    expect(declineMessage(err("billing.errors.declined", "declined"), t)).toBe(
      "DECLINED",
    );
    expect(
      declineMessage(err("billing.errors.authenticationRequired"), t),
    ).toBe("AUTH_REQUIRED");
  });

  it("falls back to the error message, then the generic copy", () => {
    expect(declineMessage(new Error("boom"), t)).toBe("boom");
    expect(declineMessage({}, t)).toBe("FAILED");
  });
});
